cloud/aws/services/ec2/interface.go

```diff
@@ -0,0 +1,22 @@
+// Copyright © 2018 The Kubernetes Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ec2
+
+import (
+	providerconfigv1 "sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+type Interface interface {
```

Comment:
I avoided defining an interface here because this package can't know how the clients are going to be using the structs it defines.

For example, this interface will grow large and if we're using it in say, the machine actuator, the interface is going to be significantly larger than it needs to be.

Reply:
I see. The main problem I was facing, and why I moved it in here was because things can get out of sync frequently at this stage. I can change it back if required, or we can refactor later as things grow.

Reply:
it's more about being very specific about what the actuator needs. If we define the interface here then we're saying "all things that use this service will need all of these methods", but that is not true since the machine actuator and the cluster actuator are using different pieces of the same struct.

Reply:
I see what you mean. Would it make more sense then to define 2 structs/interfaces inside the ec2 package? If not, I'll revert it back.

Reply:
That's still defining the interface with the library which is my main concern. In this case because there is currently one client it will be ok for a while, but it will be annoying for other variants that want to reuse portions of this logic. Not defining the interface also allows us to define only one ec2 struct that clients can scope down themselves and thus we have less boilerplate to maintain. (n services, n interfaces)

Overall, I think the point I want to make is that the client knows how they want to use the methods we provide, the  library providing those methods does not.